File: src/main/kotlin/np/com/sudanchapagain/localElectionReport/census/Census.kt
package np.com.sudanchapagain.localElectionReport.census

import org.apache.poi.ss.usermodel.Row
import org.apache.poi.xssf.usermodel.XSSFWorkbook
import java.io.FileInputStream
import java.sql.Connection
import java.sql.DriverManager

data class CensusRow(
    val province: String,
    val district: String,
    val locality: String,
    val sex: String,
    val casteEthnicity: String,
    val value: Double
)

fun censusWrangle(filePath: String): List<CensusRow> {
    FileInputStream(filePath).use { fis ->
        val wb = XSSFWorkbook(fis)
        val sheet = wb.getSheetAt(0)

        fun Row.str(colIdx: Int): String? {
            val cell = this.getCell(colIdx) ?: return null
            val s = when (cell.cellType) {
                org.apache.poi.ss.usermodel.CellType.STRING -> cell.stringCellValue
                org.apache.poi.ss.usermodel.CellType.NUMERIC -> {
                    val d = cell.numericCellValue
                    val asLong = d.toLong()
                    if (d == asLong.toDouble()) asLong.toString() else d.toString()
                }

                org.apache.poi.ss.usermodel.CellType.BOOLEAN -> cell.booleanCellValue.toString()
                org.apache.poi.ss.usermodel.CellType.FORMULA -> {
                    when (cell.cachedFormulaResultType) {
                        org.apache.poi.ss.usermodel.CellType.STRING -> cell.stringCellValue
                        org.apache.poi.ss.usermodel.CellType.NUMERIC -> {
                            val d = cell.numericCellValue
                            val asLong = d.toLong()
                            if (d == asLong.toDouble()) asLong.toString() else d.toString()
                        }

                        org.apache.poi.ss.usermodel.CellType.BOOLEAN -> cell.booleanCellValue.toString()
                        else -> cell.toString()
                    }
                }

                else -> cell.toString()
            }?.trim()
            return if (s.isNullOrEmpty()) null else s
        }

        fun Row.num(colIdx: Int): Double? {
            val cell = this.getCell(colIdx) ?: return null
            return when (cell.cellType) {
                org.apache.poi.ss.usermodel.CellType.NUMERIC -> cell.numericCellValue
                org.apache.poi.ss.usermodel.CellType.STRING -> cell.stringCellValue.toDoubleOrNull()
                else -> null
            }
        }

        val columnPROVINCE = 1
        val columnDISTRICT = 2
        val columnLOCAL = 3
        val columnSEX = 4
        val columnCASTE = 5
        val columnVALUE = 6

        val startRow = 35
        val lastRow = sheet.lastRowNum

        data class Raw(
            val rowNum: Int,
            val province: String?,
            val district: String?,
            val locality: String?,
            val sex: String?,
            val caste: String?,
            val value: Double?
        )

        val raws = mutableListOf<Raw>()
        for (r in startRow..lastRow) {
            val row = sheet.getRow(r) ?: continue
            val province = row.str(columnPROVINCE)
            val district = row.str(columnDISTRICT)
            val locality = row.str(columnLOCAL)
            val sex = row.str(columnSEX)
            val caste = row.str(columnCASTE)
            val value = row.num(columnVALUE)

            val isAllNull =
                province == null && district == null && locality == null && sex == null && caste == null && value == null
            if (isAllNull) continue

            raws += Raw(r, province, district, locality, sex, caste, value)
        }

        var currentProvince: String? = null
        val provFilled = raws.map { raw ->
            if (raw.province != null) currentProvince = raw.province
            raw.copy(province = currentProvince)
        }

        var currentDistrict: String? = null
        var lastProvinceForDistrict: String? = null
        val distFilled = provFilled.map { raw ->
            if (raw.province != lastProvinceForDistrict) {
                currentDistrict = null
                lastProvinceForDistrict = raw.province
            }
            if (raw.district != null) currentDistrict = raw.district
            raw.copy(district = currentDistrict)
        }

        var currentLocal: String? = null
        var lastProvinceForLocal: String? = null
        var lastDistrictForLocal: String? = null
        val localFilled = distFilled.map { raw ->
            if (raw.province != lastProvinceForLocal || raw.district != lastDistrictForLocal) {
                currentLocal = null
                lastProvinceForLocal = raw.province
                lastDistrictForLocal = raw.district
            }
            if (raw.locality != null) currentLocal = raw.locality
            raw.copy(locality = currentLocal)
        }

        val withGeo = localFilled.filter { it.province != null && it.district != null && it.locality != null }

        var currentSex: String? = null
        var lastLocalKey: String? = null
        val sexFilled = withGeo.map { raw ->
            val key = "${raw.province}|${raw.district}|${raw.locality}"
            if (key != lastLocalKey) {
                currentSex = null
                lastLocalKey = key
            }
            if (raw.sex != null) currentSex = raw.sex
            raw.copy(sex = currentSex)
        }

        val finalized = sexFilled.filter { it.sex != null && it.caste != null && it.value != null }
            .filter { it.locality?.trim()?.equals("INSTITUTIONAL", ignoreCase = true) != true }.map {
                CensusRow(
                    province = it.province!!,
                    district = it.district!!,
                    locality = it.locality!!,
                    sex = it.sex!!,
                    casteEthnicity = it.caste!!,
                    value = it.value!!
                )
            }

        wb.close()
        return finalized
    }
}


fun censusDataToSqlite(censusXlsxPath: String) {
    val sqlitePath: String = "res/census.db"
    val tableName = "census"
    val rows = censusWrangle(censusXlsxPath)
    val url = "jdbc:sqlite:$sqlitePath"
    DriverManager.getConnection(url).use { conn ->
        conn.autoCommit = false
        ensureTable(conn, tableName)
        clearTable(conn, tableName)
        insertRows(conn, tableName, rows)
        conn.commit()
    }
}

private fun ensureTable(conn: Connection, table: String) {
    val sql = """
        CREATE TABLE IF NOT EXISTS $table (
            province TEXT NOT NULL,
            district TEXT NOT NULL,
            locality TEXT NOT NULL,
            sex TEXT NOT NULL,
            caste_ethnicity TEXT NOT NULL,
            value REAL NOT NULL
        );
    """.trimIndent()
    conn.createStatement().use { it.execute(sql) }
}

private fun clearTable(conn: Connection, table: String) {
    conn.createStatement().use { it.execute("DELETE FROM $table") }
}

private fun insertRows(conn: Connection, table: String, rows: List<CensusRow>) {
    val sql = "INSERT INTO $table (province, district, locality, sex, caste_ethnicity, value) VALUES (?, ?, ?, ?, ?, ?)"
    conn.prepareStatement(sql).use { ps ->
        for (r in rows) {
            ps.setString(1, r.province)
            ps.setString(2, r.district)
            ps.setString(3, r.locality)
            ps.setString(4, r.sex)
            ps.setString(5, r.casteEthnicity)
            ps.setDouble(6, r.value)
            ps.addBatch()
        }
        ps.executeBatch()
    }
}

File: data/generate/census/census.py
import csv
from pathlib import Path
from openpyxl import load_workbook

ROOT = Path(__file__).resolve().parent
RAW = ROOT / "raw"
CENSUS_XLSX = RAW / "census.xlsx"
OUT_DIR = ROOT.parent.parent.parent / "generated-data"
OUT_CSV = OUT_DIR / "census.csv"

COLUMN_PROVINCE = 1
COLUMN_DISTRICT = 2
COLUMN_LOCAL = 3
COLUMN_SEX = 4
COLUMN_CASTE = 5
COLUMN_VALUE = 6
START_ROW = 35


def cell_str(ws, row_idx0, col_idx0):
    cell = ws.cell(row=row_idx0 + 1, column=col_idx0 + 1)
    v = cell.value

    if v is None:
        return None

    if isinstance(v, str):
        s = v.strip()
        return s if s != "" else None

    if isinstance(v, (int, float)):
        if isinstance(v, float) and v.is_integer():
            return str(int(v))
        return str(v)

    if isinstance(v, bool):
        return str(v)

    s = str(v).strip()

    return s if s != "" else None


def cell_num(ws, row_idx0, col_idx0):
    cell = ws.cell(row=row_idx0 + 1, column=col_idx0 + 1)
    v = cell.value

    if v is None:
        return None

    if isinstance(v, (int, float)):
        return float(v)

    if isinstance(v, str):
        try:
            return float(v)
        except Exception:
            return None

    return None


def wrangle(census_xlsx_path: Path):
    wb = load_workbook(filename=str(census_xlsx_path), data_only=True)
    sheet = wb.worksheets[0]

    last_row0 = sheet.max_row - 1
    raws = []

    for r in range(START_ROW, last_row0 + 1):
        province = cell_str(sheet, r, COLUMN_PROVINCE)
        district = cell_str(sheet, r, COLUMN_DISTRICT)
        locality = cell_str(sheet, r, COLUMN_LOCAL)
        sex = cell_str(sheet, r, COLUMN_SEX)
        caste = cell_str(sheet, r, COLUMN_CASTE)
        value = cell_num(sheet, r, COLUMN_VALUE)

        if all(v is None for v in [province, district, locality, sex, caste, value]):
            continue

        raws.append(
            {
                "rownum": r,
                "province": province,
                "district": district,
                "locality": locality,
                "sex": sex,
                "caste_ethnicity": caste,
                "value": value,
            }
        )

    current_province = None
    for i, raw in enumerate(raws):
        if raw["province"] is not None:
            current_province = raw["province"]

        raws[i]["province"] = current_province

    current_district = None
    last_province_for_district = None
    for i, raw in enumerate(raws):
        if raw["province"] != last_province_for_district:
            current_district = None
            last_province_for_district = raw["province"]

        if raw["district"] is not None:
            current_district = raw["district"]

        raws[i]["district"] = current_district

    current_local = None
    last_province_for_local = None
    last_district_for_local = None

    for i, raw in enumerate(raws):
        if (
            raw["province"] != last_province_for_local
            or raw["district"] != last_district_for_local
        ):
            current_local = None
            last_province_for_local = raw["province"]
            last_district_for_local = raw["district"]

        if raw["locality"] is not None:
            current_local = raw["locality"]

        raws[i]["locality"] = current_local

    with_geo = [r for r in raws if r["province"] and r["district"] and r["locality"]]
    current_sex = None
    last_local_key = None

    for raw in with_geo:
        key = f"{raw['province']}|{raw['district']}|{raw['locality']}"

        if key != last_local_key:
            current_sex = None
            last_local_key = key

        if raw["sex"] is not None:
            current_sex = raw["sex"]

        raw["sex"] = current_sex

    finalized = [
        r
        for r in with_geo
        if r["sex"] is not None
        and r["caste_ethnicity"] is not None
        and r["value"] is not None
        and not (
            isinstance(r["locality"], str)
            and r["locality"].strip().lower() == "institutional"
        )
    ]

    wb.close()
    return finalized


if __name__ == "__main__":
    if not CENSUS_XLSX.exists():
        raise SystemExit(f"census xlsx not found at {CENSUS_XLSX}")

    rows = wrangle(CENSUS_XLSX)
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    rows_csv = [{k: v for k, v in r.items() if k != "rownum"} for r in rows]

    with OUT_CSV.open("w", newline="", encoding="utf-8") as fh:
        writer = csv.DictWriter(
            fh,
            fieldnames=[
                "province",
                "district",
                "locality",
                "sex",
                "caste_ethnicity",
                "value",
            ],
        )

        writer.writeheader()
        writer.writerows(rows_csv)
